cpp/src/arrow/adapters/orc/jni/jni_wrapper.cpp

```diff
@@ -0,0 +1,243 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/adapters/orc/adapter.h>
+#include <arrow/array.h>
+#include <arrow/buffer.h>
+#include <arrow/io/api.h>
+#include <arrow/ipc/api.h>
+#include <arrow/util/logging.h>
+#include <cassert>
+#include <string>
+
+#include "org_apache_arrow_adapter_orc_OrcMemoryJniWrapper.h"
+#include "org_apache_arrow_adapter_orc_OrcReaderJniWrapper.h"
+#include "org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper.h"
+
+using ORCFileReader = arrow::adapters::orc::ORCFileReader;
+using RecordBatchReader = arrow::RecordBatchReader;
+
+void ThrowJavaIOException(JNIEnv* env, const std::exception& e) {
+  jclass ioExceptionClass = env->FindClass("java/io/IOException");
+  if (ioExceptionClass != nullptr) {
+    if (env->ThrowNew(ioExceptionClass, e.what())) {
+      // Failed to new IOException. This means another error has occurred in Java
+      // We just propagate this error to caller by doing nothing.
+      ARROW_LOG(ERROR) << "Error occurred when throwing IOException";
+    }
+  } else {
+    ARROW_LOG(ERROR) << "Error occurred when getting IOException class";
+  }
+}
+
+void ThrowJavaException(JNIEnv* env, const std::string& message) {
+  jclass exception = env->FindClass("java/lang/Exception");
+  if (exception != nullptr) {
+    env->ThrowNew(exception, message.c_str());
+  } else {
+    throw std::runtime_error("Can't find java/lang/Exception class");
+  }
+}
+
+jfieldID GetFieldId(JNIEnv* env, jclass this_class, const std::string& sig) {
+  jfieldID ret = env->GetFieldID(this_class, sig.c_str(), "J");
+  if (ret == nullptr) {
+    ThrowJavaException(env, "Unable to get java class field: " + sig);
+  }
+
+  return ret;
+}
+
+std::unique_ptr<ORCFileReader>* GetNativeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj, GetFieldId(env, env->GetObjectClass(this_obj), "nativeReaderAddress"));
+  return reinterpret_cast<std::unique_ptr<ORCFileReader>*>(reader);
+}
+
+std::shared_ptr<RecordBatchReader>* GetStripeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj,
+      GetFieldId(env, env->GetObjectClass(this_obj), "nativeStripeReaderAddress"));
+  return reinterpret_cast<std::shared_ptr<RecordBatchReader>*>(reader);
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_open(
+    JNIEnv* env, jobject this_obj, jstring file_path) {
+  std::shared_ptr<arrow::io::ReadableFile> in_file;
+  const char* str = env->GetStringUTFChars(file_path, nullptr);
+  std::string path(str);
+  env->ReleaseStringUTFChars(file_path, str);
+
+  arrow::Status ret;
+  if (path.find("hdfs://") == 0) {
+    return false;
+  } else {
+    ret = arrow::io::ReadableFile::Open(path, &in_file);
+  }
+
+  if (ret.ok()) {
+    auto reader = new std::unique_ptr<ORCFileReader>();
```

Comment:
see note about memory management below (i.e. use a unique pointer and release it when you are ready to assign it.